crypto/threads_pthread.c

```diff
@@ -42,12 +42,401 @@
+     * left over, so that readers can start working on
+     * one that isn't yet being waited on
+     */
+    while (ilock->group_count - ilock->writers_alloced < 2) {
+        /*
+         * we have to wait for one to be free
+         */
+        pthread_cond_wait(&ilock->alloc_signal, &ilock->alloc_lock);
+    }
+
+    current_idx = ilock->current_alloc_idx;
+    /*
+     * Allocate the qp
+     */
+    ilock->writers_alloced++;
+
+    /*
+     * increment the allocation index
+     */
+    ilock->current_alloc_idx =
+        (ilock->current_alloc_idx + 1) % ilock->group_count;
+
+    /*
+     * get and insert a new id
+     */
+    new_id = __atomic_fetch_add(&ilock->id_ctr, 1, __ATOMIC_SEQ_CST);
+
+    new_id = VAL_ID(new_id);
+    __atomic_and_fetch(&ilock->qp_group[current_idx].users, ID_MASK,
+                       __ATOMIC_SEQ_CST);
+    __atomic_or_fetch(&ilock->qp_group[current_idx].users, new_id,
+                      __ATOMIC_SEQ_CST);
+
+    /*
+     * update the reader index to be the prior qp
+     */
+    __atomic_store_n(&ilock->reader_idx, ilock->current_alloc_idx,
+                     __ATOMIC_SEQ_CST);
+
+    /*
+     * wake up any waiters
+     */
+    pthread_cond_signal(&ilock->alloc_signal);
+    pthread_mutex_unlock(&ilock->alloc_lock);
+    return (struct rcu_qp *)&ilock->qp_group[current_idx];
+}
+
+static void retire_qp(struct rcu_lock_internal *ilock,
+                      struct rcu_qp *qp)
+{
+    pthread_mutex_lock(&ilock->alloc_lock);
+    ilock->writers_alloced--;
+    pthread_cond_signal(&ilock->alloc_signal);
+    pthread_mutex_unlock(&ilock->alloc_lock);
+}
+
+static struct rcu_qp *allocate_new_qp_group(struct rcu_lock_internal *ilock,
+                                            int count)
+{
+    struct rcu_qp *new =
+        OPENSSL_zalloc(sizeof(struct rcu_qp) * count);
+
+    ilock->group_count = count;
+    return new;
+}
+
+/*
+ * Public WRITE api
+ */
+void CRYPTO_THREAD_rcu_write_lock(CRYPTO_RCU_LOCK *lock)
+{
+    struct rcu_lock_internal *ilock = (struct rcu_lock_internal *)lock;
+
+    pthread_mutex_lock(&ilock->write_lock);
+}
+
+/*
+ * Public READ api
+ */
+void CRYPTO_THREAD_rcu_write_unlock(CRYPTO_RCU_LOCK *lock)
+{
+    struct rcu_lock_internal *ilock = (struct rcu_lock_internal *)lock;
+
+    pthread_mutex_unlock(&ilock->write_lock);
+}
+
+/*
+ * Public WRITE api for synchronization
+ */
+void CRYPTO_THREAD_synchronize_rcu(CRYPTO_RCU_LOCK *lock)
+{
+    struct rcu_qp *qp;
+    struct rcu_lock_internal *ilock;
+    uint64_t count;
+    struct rcu_cb_item *cb_items, *tmpcb;
+
+    /*
+     * before we do anything else, lets grab the cb list
+     */
+    ilock = (struct rcu_lock_internal *)lock;
+    cb_items = __atomic_exchange_n(&ilock->cb_items, NULL, __ATOMIC_SEQ_CST);
+
+    qp = update_qp(ilock);
+
+    /*
+     * wait for the reader count to reach zero
+     */
+    do {
+        count = __atomic_load_n(&qp->users, __ATOMIC_SEQ_CST);
+    } while (READER_COUNT(count) != 0);
```

Comment:
So....a busy wait?? We assume then that reading is fast?

Reply:
ossl_synchronize_rcu is a write site operation.  It implements the delay needed to wait for any readers of shared data to complete.  It could have its overhead reduced via the use of pthread_conditional, but the nature of atomics on the read side make its very easy to run into deadlocks in the signaling.